### PR TITLE
Refactor remaining local exceptions in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR renames the following exceptions:
CloseConnectionException -> ConnectionClosedByPeerException
ConnectionClosedManuallyException -> ConnectionClosedLocallyException (and drops Graceful)
Timeout -> ConnectionTimeoutException

The new InvocationTimeout exception is now an "abstract" System.Exception with two derived classes: InvocationTimeoutException and InvocationCanceledException.

ConnectTimeoutException now derives from ConnectFailedException.

### CODE REVIEW

File: csharp/test/Ice/metrics/AllTests.cs [hunk 1]
```diff
@@ -627,9 +627,8 @@ public static IMetricsPrx allTests(TestHelper helper, CommunicatorObserver obsv)
                 }
                 Thread.Sleep(10);
             }
-            test(cm1.Failures == 2 && sm1.Failures >= 2);
+            test(cm1.Failures == 1 && sm1.Failures >= 2);
```
Comment (bernardnormier) [inline csharp/test/Ice/metrics/AllTests.cs]: It's unclear why the exception renaming/refactoring changed this number.

File: csharp/test/Ice/metrics/AllTests.cs [hunk 2]
```diff
@@ -612,7 +612,7 @@ public static IMetricsPrx allTests(TestHelper helper, CommunicatorObserver obsv)
                 metrics.Clone(connectionTimeout: 500).opByteS(new byte[10000000]);
```
Comment (pepone) [inline csharp/test/Ice/metrics/AllTests.cs]: not clear why we do this clone if we are setting the same timeout, isn't this returning the same proxy?

File: csharp/test/Ice/metrics/AllTests.cs (hunk 1 above)
Comment (pepone) [reply csharp/test/Ice/metrics/AllTests.cs]: Seems like the invocation was causing a retry before, but not anymore.

File: csharp/test/Ice/metrics/AllTests.cs (hunk 2 above)
Comment (pepone) [reply csharp/test/Ice/metrics/AllTests.cs]: nvm the previous call didn't assign the proxy

**Comment by bernardnormier** [comment]
Add new exception class ConnectionIdleTimeout (as suggesed by Benoit), renamed InvocationTimeoutException -> System.TimeoutException, InvocationCanceledException -> System.OperationCanceledException, fixed tests.

File: csharp/src/Ice/Connection.cs [hunk 3]
```diff
@@ -677,7 +677,7 @@ public void AsyncRequestCanceled(OutgoingAsyncBase outAsync, System.Exception ex
                         _asyncRequests.Remove(o.RequestId);
                     }
 
-                    if (ex is ConnectionTimeoutException)
+                    if (ex is ConnectionIdleException)
```
Comment (bentoi) [inline csharp/src/Ice/Connection.cs]: This isn't correct. The ConnectionTimeoutException check here was to support `invocationTimeout = -2` which is supposed to close the connection on invocation timeout. You changed the invocation timeout cancellation to now use TimeoutException for both `invocationTimeout = -2` and `!= -2` so there's no way to distinguish anymore between these 2 cases. We should either remove support for invocationTimeout = -2 and this `if()` block or allow the code to distinguish between the 2 reasons for invocation timeouts.

File: csharp/src/Ice/Connection.cs [hunk 4]
```diff
@@ -710,7 +710,7 @@ public void AsyncRequestCanceled(OutgoingAsyncBase outAsync, System.Exception ex
                     {
                         if (kvp.Value == outAsync)
                         {
-                            if (ex is ConnectionTimeoutException)
+                            if (ex is ConnectionIdleException)
```
Comment (bentoi) [inline csharp/src/Ice/Connection.cs]: Same issue there, ex is never ConnectionIdleException here.

File: csharp/src/Ice/OutgoingAsync.cs [hunk 5]
```diff
@@ -593,17 +595,9 @@ protected override bool ResponseImpl(bool userThread, bool ok, bool invoke)
...
-            if (Proxy.IceReference.GetInvocationTimeout() == -2)
-            {
-                Cancel(new Ice.ConnectionTimeoutException());
-            }
-            else
-            {
-                Cancel(new Ice.InvocationTimeoutException());
-            }
-        }
+            => Cancel(new TimeoutException($"invocation on remote Ice object `{Proxy.Identity}' timed out"));
```
Comment (bentoi) [inline csharp/src/Ice/OutgoingAsync.cs]: Using the same exception effectively removes the support for `invocationTimeout = -2`.

File: csharp/test/Ice/retry/AllTests.cs [hunk 6]
```diff
@@ -240,26 +241,26 @@ static public Test.IRetryPrx
...
                     Instrumentation.testRetryCount(-1);
                 }
                 if (retry1.GetConnection() != null)
                 {
-                    // The timeout might occur on connection establishment or because of the sleep. What's
-                    // important here is to make sure there are 4 retries and that no calls succeed to
-                    // ensure retries with the old connection timeout semantics work.
+                    // The old invocationTimeout = -2 behaves like any timeout now, with no retry
+                    // TODO: should we remove this entirely? can the connection establishment timeout and trigger a
+                    // retry?
```
Comment (bentoi) [inline csharp/test/Ice/retry/AllTests.cs]: I don't see the point of introducing new semantics for invocationTimeout = -2, I think we should just remove support for it.

File: csharp/src/Ice/Connection.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs]: Fixed by removing `invocationTimeout = -2`. It's no longer a valid value.

File: csharp/src/Ice/Connection.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/Connection.cs]: Fixed (removed)

File: csharp/test/Ice/retry/AllTests.cs (hunk 6 above)
Comment (bernardnormier) [reply csharp/test/Ice/retry/AllTests.cs]: Done in latest commit

**Pull request merged**